integration/nwo/common/pkcs11/pkcs11.go

```diff
@@ -0,0 +1,244 @@
+	if opts.Library == "" {
+		return nil, fmt.Errorf("pkcs11: library path not provided")
+	}
+
+	ctx := pkcs11lib.New(opts.Library)
+	if ctx == nil {
+		return nil, fmt.Errorf("pkcs11: instantiation failed for %s", opts.Library)
+	}
+	defer ctx.Finalize()
+	if err := ctx.Initialize(); err != nil {
+		return nil, fmt.Errorf("pkcs11: initialization failed: %v", err)
+	}
+
+	slots, err := ctx.GetSlotList(false)
+	if err != nil {
+		return nil, fmt.Errorf("pkcs11: failed to get slot list: %v", err)
+	}
+
+	var tokens []pkcs11lib.TokenInfo
+	for _, s := range slots {
+		info, err := ctx.GetTokenInfo(s)
+		if err != nil {
+			fmt.Printf("pkcs11: failed to get token info for slot %d: %v", s, err)
+			continue
+		}
+		tokens = append(tokens, info)
+	}
+	return tokens, nil
+}
+
+// CheckToken checks whether a token is available that matches the given label.
+// If the token does not exist, it is created.
+func CheckToken(opts *config.PKCS11) error {
+	if opts.Library == "" {
+		return fmt.Errorf("pkcs11: library path not provided")
+	}
+
+	ctx := pkcs11lib.New(opts.Library)
+	if ctx == nil {
+		return fmt.Errorf("pkcs11: instantiation failed for %s", opts.Library)
+	}
+	defer ctx.Finalize()
+	if err := ctx.Initialize(); err != nil && !strings.Contains(err.Error(), "CKR_CRYPTOKI_ALREADY_INITIALIZED") {
+		return fmt.Errorf("pkcs11: initialization failed: %v", err)
+	}
+
+	slots, err := ctx.GetSlotList(false)
+	if err != nil {
+		return fmt.Errorf("pkcs11: failed to get slot list: %v", err)
+	}
+
+	for _, s := range slots {
+		info, err := ctx.GetTokenInfo(s)
+		if err != nil {
+			fmt.Printf("pkcs11: failed to get token info for slot %d: %v", s, err)
+			continue
+		}
+		if info.Label == opts.Label {
+			fmt.Printf("pkcs11: found token with label %s on slot %d \n", opts.Label, s)
+			return nil
+		}
+	}
+
+	fmt.Printf("pkcs11: no token with label %s found, create one with PIN %s", opts.Label, opts.Pin)
+
+	// create token
+	if err := ctx.InitToken(0, opts.Pin, opts.Label); err != nil {
+		return fmt.Errorf("pkcs11: failed to initialize token: %v", err)
+	}
+
+	slots, err = ctx.GetSlotList(true)
+	if err != nil {
+		return errors.Wrap(err, "pkcs11: get slot list")
+	}
+	var slot uint
+	for _, s := range slots {
+		info, err := ctx.GetTokenInfo(s)
+		if err != nil || opts.Label != info.Label {
+			continue
+		}
+		slot = s
+		break
+	}
+
+	sess, err := ctx.OpenSession(slot, pkcs11lib.CKF_SERIAL_SESSION|pkcs11lib.CKF_RW_SESSION)
+	if err != nil {
+		return errors.Wrap(err, "pkcs11: open session")
+	}
+	defer ctx.CloseSession(sess)
+	if err := ctx.Login(sess, pkcs11lib.CKU_USER, opts.Pin); err != nil {
+		return errors.Wrap(err, "pkcs11: login")
+	}
+	if err := ctx.InitPIN(sess, opts.Pin); err != nil {
+		return errors.Wrap(err, "pkcs11: init pin")
+	}
+
+	return nil
+}
+
+// FindPKCS11Lib attempts to find the PKCS11 library based on the given configuration
+func FindPKCS11Lib() (lib, pin, label string) {
+	if lib = os.Getenv("PKCS11_LIB"); lib == "" {
```

Comment:
Where is this ENV var set? Is it documented somewhere?

Reply:
I copied this function from Fabric. We don't have documentation about NWO yet.
We need a dedicated PR for this documentation.

Reply:
NWO generates the configuration files with the right paths directly.